### PR TITLE
build(electron): asar

### PR DESCRIPTION
**Pull request by pengx17**
Due to restrictions on how Electron package works, the `node_modules` should not be hoisted and not to use s/h-links at all. This is why we need to have two separate installs for electron and non-electron packages in the build.

Tested via the following script

```bash
#!/bin/bash

echo "step 1: clean up"
find . -name "node_modules" -prune -exec rm -rf '{}' +
# git clean -dfX
build_type=canary

echo "step 2: install web dependencies"
# firstly, build web static
PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 SENTRYCLI_SKIP_DOWNLOAD=1 PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 HUSKY=1 yarn

echo "step 3: generate assets"
BUILD_TYPE="$build_type" yarn workspace @affine/electron generate-assets

# cleanup node_modules
find . -name "node_modules" -prune -exec rm -rf '{}' +

echo "step 4: install electron dependencies"
# install electron deps
yarn config set nmHoistingLimits workspaces
yarn config set enableScripts false
yarn config set nmMode classic
PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 HUSKY=0 yarn workspaces focus @affine/electron @affine/monorepo

echo "step 5: build native"
# build native
yarn workspace @affine/native build
yarn workspace @affine/storage build

echo "step 6: build electron"
# build electron
yarn workspace @affine/electron build

echo "step 7: package electron"
# package
SKIP_GENERATE_ASSETS=1 BUILD_TYPE="$build_type" HOIST_NODE_MODULES=1 yarn workspace @affine/electron package
```

### CODE REVIEW

**Comment by pengx17** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #4965** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/4965?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/4965?utm_source=stack-comment).

File: yarn.lock [hunk 1]
```diff
@@ -19293,6 +19304,19 @@ __metadata:
   languageName: node
   linkType: hard
 
+"electron@npm:^27.1.0":
```
Comment (Brooooooklyn) [inline yarn.lock:19307]: there is also a `27.0.0` electron in lockfile, you should dedupe it

File: .github/actions/setup-node/action.yml [hunk 2]
```diff
@@ -49,9 +49,9 @@ runs:
         cache: 'yarn'
 
     - name: Set nmMode
-      if: ${{ inputs.hard-link-nm == 'true' }}
+      if: ${{ inputs.hard-link-nm == 'false' }}
       shell: bash
-      run: yarn config set nmMode hardlinks-local
+      run: yarn config set nmMode classic
```
Comment (Brooooooklyn) [inline .github/actions/setup-node/action.yml:54]: you should set it false in the electron build step rather than modify the `actions/setup-node`

Comment (pengx17) [reply .github/actions/setup-node/action.yml:54]: `nmMode`'s default value in our `.yarnrc` is already `hardlinks-local`, so existing `hard-link-nm == 'true'` in fact does nothing.
https://github.com/toeverything/AFFiNE/blob/29d086652f8b3f02044f47b0ba4f186f1298a414/.yarnrc.yml#L5

**Review by Brooooooklyn — APPROVED** [review]

**Review by Brooooooklyn — DISMISSED** [review]

**Review by Brooooooklyn — APPROVED** [review]

**Pull request merged**